File: 1-100/69-sqrt/sqrt.cpp
class Solution {
   public:
    int mySqrt(int x) {
        if (x <= 1) return x;

        int left = 1, right = x / 2;
        int mid;

        while (true) {
            mid = (left + right) / 2;

            if (mid > x / mid)
                right = mid - 1;
            else if ((mid + 1) > x / (mid + 1))
                return mid;
            else
                left = mid + 1;
        }
    }
};

File: 1-100/61-70/69-sqrt/sqrt.py
class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """

        if not x:
            return x

        original = x
        nextx = (x + original / x) / 2.0
        while x - nextx > 0.000001:
            x = nextx
            nextx = (x + original / x) / 2.0

        return int(nextx)